src/mem/Allocator.h

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 1996-2022 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_SRC_MEM_ALLOCATOR_H
+#define SQUID_SRC_MEM_ALLOCATOR_H
+
+#include "mem/forward.h"
+
+namespace Mem
+{
+
+/// An interface for memory allocators that deal with fixed-size objects.
+/// Allocators may optimize repeated de/allocations using memory pools.
+class Allocator
+{
+public:
+    Allocator(char const *aLabel) : label(aLabel) {}
+    virtual ~Allocator() {}
+
+    // TODO make this method const
+    /**
+     * fill the given object with statistical data about pool
+     * \returns Number of objects in use, ie. allocated.
+     */
+    virtual int getStats(MemPoolStats *, int accumulate = 0) = 0;
+
+    virtual MemPoolMeter const &getMeter() const = 0;
+
+    /// allocate one element from the pool
+    virtual void *alloc() = 0;
+
+    /// free a element allocated by Allocator::alloc()
+    virtual void freeOne(void *) = 0;
+
+    /// get a display label for objects in this pool
+    virtual char const *objectType() const { return label; }
+
+    /// the size (in bytes) of objects managed by this allocator
+    virtual size_t objectSize() const = 0;
+
+    /// the number of objects currently allocated
+    virtual int getInUseCount() = 0;
+
+    /// \see doZero
+    void zeroBlocks(bool doIt) {doZero = doIt;}
+
+    int inUseCount() { return getInUseCount(); } // XXX: drop redundant?
+
+    /**
+     * Allows you tune chunk size of pooling. Objects are allocated in chunks
+     * instead of individually. This conserves memory, reduces fragmentation.
+     * Because of that memory can be freed also only in chunks. Therefore
+     * there is tradeoff between memory conservation due to chunking and free
+     * memory fragmentation.
+     *
+     * \note  As a general guideline, increase chunk size only for pools that keep
+     *        very many items for relatively long time.
+     */
```

Comment:
Huh? this makes me think you do not understand the purpose of a virtual is. Whether the pool has an internal concept of chunk or not,  the interface and the caller do and they need to communicate about chunks.

Reply:
> Whether the pool has an internal concept of chunk or not, the interface and the caller do and they need to communicate about chunks.

This interface and the callers using it are currently broken (because some of the allocators do not support a notion of a "chunk"). This method does not belong to the root hierarchy class. That base class is supposed to define what is _common_ among derived classes. It is not a collection of all the things some child classes can do.